README.md
# redmoon

[![version](https://img.shields.io/npm/v/redmoon.svg) ![download](https://img.shields.io/npm/dm/redmoon.svg)](https://www.npmjs.com/package/redmoon)

Asynchronous broker for multi-source data collection using redis.

If redmoon.load is called, and it returned to find the data in the cache, if cache data does not exist, call the specified function(subscriber) to subscribe.

[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)

## Usage

```javascript

// setting
var redmoon = require('redmoon').create(config)

// event handle
redmoon
	.on('timeout', function(keyword) {
		console.error('redmoon timeout', keyword)
	})
	.on('error', function(err) {
		console.error('redmoon error', err)
	});

// express
app.get('/search/:key/:page/:limit?', function (req, res) {
  redmoon.load(function (err, items, meta) {
    res.json({
      meta: meta,
      items: items
    })
  }, req.params.key, req.params.page, req.params.limit)
})

// collector
// The first event is no meta data in moon.
redmoon
	.subscribe(subscriber)
	.garbage.start(3600)	// It deletes the data collected before to 60 minutes.

var subscriber = function (moon) {
	redmoon.atomic(moon.uuid, function(cb) {
		process(cb, moon)
	}, function(err, data) {
		complete(moon, data)
	})
}

var process = function (cb, moon) {
	// call cb when complete load data
	cb(err, {
		meta: {},
		items: {}
	})
}

var complete = function(moon, data) {
	redmoon.add(function(err) {
		redmoon.trigger(moon.topic)
	}, moon, data.meta, data.items)
}


// moon
// {
// 	uuid: uuid,
// 	key: key,
// 	topic: topic,
// 	meta: meta	// if first event then it is undefined
// }

```

## Methods

### redmoon.connect(config)

Connect redis.

### redmoon.events()

Initialize redmoon events.
* error
* ready
* connect
* end

### redmoon.load(cb(err, items, meta), key, page, limit)

Load collected data.
If not exist data then trigger event for collection.

### redmoon.loadMeta(cb(err, metas), uuid, count)

Load meta data.

### redmoon.add(cb, moon, meta, items)

Add collected data to cache.

### redmoon.subscribe(cb(err, moon))

Set callback function for the collection event.

### redmoon.unsubscribe(cb())

Unset callback function for the collection event.

### redmoon.trigger(topic, data)

Trigger event for the specified topic.

### redmoon.atomic(uuid, process(cb), complete(err, data))

The function to prevent the duplication proccess.

If it has available resources, given `process` calls and then release the resources used by `cb` call in process.
After the release, and it executes the transfer of data in `proccess` to `complete`.

### redmoon.end()

Close the connection to the redis server.

### redmoon.truncate(unix)

Deleting cache data collected before a specified time.

### redmmon.toTopic(uuid)

It converts uuid to the topic.

### redmoon.garbage.start(offset)

Start garbage collection.
Deleting cache data collected before a offset time.

### redmoon.garbage.stop()

Stop garbage collection.

### Redmoon.uuid(key)

Make unique string.

### Redmmon.uuid()

Return unixtimestamp.

### Redmoon.create(config)

Create Redmoon instance with config.


## Release History

See the [changelog](CHANGELOG.md)

## LICENSE

redmoon is licensed under the MIT license.


CHANGELOG.md
# 1.0.2

2015-11-14

add random string to topic to permit duplicate request.

# 1.0.1

2015-11-14

- fixed bug that additional data collection.

# 1.0.0

2015-11-12

- add garbage collector.
- improve multiple request processing logic.

# 0.9.1

2015-11-09

 - add events for additional data collection.

# 0.9.0

2015-11-06

 - initial release.


test.js
// test

var config = {
  scope: 'redmoon',
  name: 'default',
  host: '127.0.0.1',
  port: 6379,
  ttl: 3600 * 12,
  timeout: 5000 // 5 sec
}

var redmoon = require('./').create(config)

redmoon.on('timeout', function (key) {
  console.log('timeout', key)
})

redmoon.subscribe(function (moon) {
  console.info('receive event for collection.', moon)

  redmoon.add(function (err) {
    redmoon.trigger(moon.topic)
  }, moon, {
    provider: 'provider'
  }, [moon.key])
})

setTimeout(function () {
  redmoon.load(function (err, result) {
    if (err) {
      return console.error('load error', err)
    }

    console.log(result)
  }, process.argv[2] || 'redmoon')
}, 500)


lib/redmoon.js
var util = require('util')
var events = require('events')
var shortid = require('shortid')
var NRP = require('node-redis-pubsub')

function Redmoon (config) {
  events.EventEmitter.call(this)

  this._client = null
  this._nrp = null

  this._timer = null

  this.config = {
    scope: 'redmoon',
    name: 'default',
    host: '127.0.0.1',
    port: 6379,
    timeout: 5000, // 5 sec
    buffer: 100, // buffer count for preloading
    ttl: 5, // max time for atomic
    interval: 60 // interval value for garbage collection
  }
  this.key = {
    atomic: [this.config.scope, 'atomic'].join(':'), // hash
    meta: [this.config.scope, 'meta'].join(':'), // hash
    cycle: [this.config.scope, 'cycle'].join(':') // set (sorted)
  }

  this.garbage = {
    start: function (offset) {
      this._timer = setInterval(function () {
        this.truncate(Redmoon.unix() - offset)
      }.bind(this), this.config.interval * 1000)
    }.bind(this),
    stop: function () {
      clearInterval(this._timer)
      this._timer = null
    }.bind(this)
  }
}

util.inherits(Redmoon, events.EventEmitter)

Redmoon.prototype.connect = function (config) {
  this.end()

  config = config || this.config

  typeof config.scope !== 'undefined' && (this.config.scope = config.scope)
  typeof config.name !== 'undefined' && (this.config.name = config.name)
  typeof config.host !== 'undefined' && (this.config.host = config.host)
  typeof config.port !== 'undefined' && (this.config.port = config.port)
  typeof config.timeout !== 'undefined' && (this.config.timeout = config.timeout)
  typeof config.buffer !== 'undefined' && (this.config.buffer = config.buffer)
  typeof config.ttl !== 'undefined' && (this.config.ttl = config.ttl)
  typeof config.interval !== 'undefined' && (this.config.interval = config.interval)

  this._client = require('redis').createClient(this.config.port, this.config.host)
  this._nrp = new NRP(this.config)

  this.events()

  return this
}

Redmoon.prototype.events = function () {
  var self = this
  var events = ['error', 'ready', 'connect', 'end']

  for (var i = 0; i < events.length; i++) {
    this._client.on(events[i], (function (evt) {
      return function () {
        var args = Array.prototype.slice.call(arguments)

        args.unshift(evt)
        self.emit.apply(self, args)
      }
    }(events[i])))
  }

  return this
}

Redmoon.prototype.load = function (cb, key, page, limit) {
  var self = this
  var uuid = Redmoon.uuid(key)

  page = page || 1
  limit = limit || 10

  // check exist search result on meta
  this.loadMeta(function (err, metas) {
    if (err) {
      return cb(err)
    }

    if (metas.length > 0) {
      var start = (page - 1) * limit
      var end = start + limit - 1

      self._client.lrange([self.config.scope, self.config.name, uuid].join(':'), start, end, function (err, items) {
        if (err) {
          return cb(err)
        }

        // migrate items
        for (var i = 0, len = items.length; i < len; i++) {
          items[i] = JSON.parse(items[i])
        }

        // migrate meta
        var meta = {}
        var loadedcount = 0
        for (var j = 1, parsed = null; j < metas.length; j += 2) {
          parsed = JSON.parse(metas[j])
          meta[parsed.provider] = parsed
          loadedcount += (parsed.page - 1) * (parsed.limit || 0)
        }

        // required collection
        if (end + self.config.buffer > loadedcount) {
          self._nrp.emit([self.config.scope, self.config.name, 'request', uuid].join(':'), {
            uuid: uuid,
            key: key,
            topic: self.toTopic(uuid),
            meta: meta
          })
        }

        cb(null, items, meta)
      })
    } else {
      var timeout = null
      var topic = self.toTopic(uuid)

      timeout = setTimeout(function () {
        self._nrp.off(topic)
        self.emit('timeout', key)

        cb('redmoon.timeout')
      }, self.config.timeout)

      self._nrp.on(topic, function (data) {
        timeout && clearTimeout(timeout)

        self._nrp.off(topic)
        self.load(cb, key, page, limit)
      })
      self._nrp.emit([self.config.scope, self.config.name, 'request', uuid].join(':'), {
        uuid: uuid,
        key: key,
        topic: topic
      })
    }
  }, uuid)

  return this
}

Redmoon.prototype.loadMeta = function (cb, uuid, count) {
  this._client.send_command('HSCAN', [this.key.meta, 0, 'MATCH', uuid + ':*', 'COUNT', count || 1000], function (err, metas) {
    cb(err, metas ? metas[1] : [])
  })

  return this
}

Redmoon.prototype.add = function (cb, moon, meta, items) {
  var uuid = Redmoon.uuid(moon.key)
  var multi = this._client.multi()
    .zadd(this.key.cycle, Redmoon.unix(), uuid)
    .hset(this.key.meta, [uuid, meta.provider || shortid.generate()].join(':'), JSON.stringify(meta))

  for (var i = 0, len = items.length; i < len; i++) {
    multi.rpush([this.config.scope, this.config.name, uuid].join(':'), JSON.stringify(items[i]))
  }

  multi.exec(function (err) {
    cb(err)
  })

  return this
}

Redmoon.prototype.subscribe = function (cb) {
  this._nrp.on([this.config.scope, this.config.name, 'request', '*'].join(':'), cb)
  return this
}

Redmoon.prototype.unsubscribe = function (cb) {
  this._nrp.off([this.config.scope, this.config.name, 'request', '*'].join(':'), cb)
  return this
}

Redmoon.prototype.trigger = function (topic, data) {
  this._nrp.emit(topic, data || {})
  return this
}

// atomic processing
Redmoon.prototype.atomic = function (uuid, process, complete) {
  // atomic
  var self = this
  var key = [this.key.atomic, uuid].join(':')

  this._client.get(key, function (err, item) {
    if (err) {
      return self.emit('error', err)
    }

    if (item) {
      // already requested
      return true
    }

    self._client.multi()
      .set(key, uuid)
      .expire(key, self.config.ttl)
      .exec(function (err) {
        if (err) {
          return self.emit('error', err)
        }

        process(function (err, data) {
          self._client.del([self.key.atomic, uuid].join(':'), function () {
            complete(err, data)
          })
        })
      })
  })
  return this
}

Redmoon.prototype.end = function () {
  if (this._client) {
    this._client.end()
    this._client = null
  }

  if (this._nrp) {
    this._nrp.end()
    this._nrp = null
  }
  return this
}

Redmoon.prototype.truncate = function (unix) {
  var self = this
  unix = unix || Redmoon.unix() - 3600 // default 1 hour

  this._client.zrangebyscore(this.key.cycle, 0, unix, function (err, items) {
    if (err) {
      return self.emit('error', err)
    }

    for (var i = 0; i < items.length; i++) {
      self.loadMeta((function (uuid) {
        return function (err, metas) {
          if (err) {
            return self.emit('error', err)
          }

          var multi = self._client.multi()

          for (var j = 0; j < metas.length; j += 2) {
            multi.hdel(self.key.meta, metas[j])
          }

          multi.del([self.config.scope, self.config.name, uuid].join(':'))
          multi.zrem(self.key.cycle, uuid)
          multi.exec(function (err, replies) {
            if (err) {
              return self.emit('error', err)
            }
          })
        }
      })(items[i]), items[i])
    }
  })
  return this
}

Redmoon.prototype.toTopic = function (uuid) {
  return [this.config.name, 'result', uuid, shortid.generate()].join(':')
}

Redmoon.uuid = function (key) {
  return encodeURIComponent(key.replace(/:/gi, '-'))
}

Redmoon.unix = function () {
  return Math.floor((+new Date()) / 1000)
}

Redmoon.create = function (config) {
  return (new Redmoon()).connect(config || {})
}

module.exports = Redmoon
